include/pybind11/pybind11.h

```diff
@@ -1110,22 +1147,84 @@ template <typename Type, typename... Extra> iterator make_iterator(Type &value,
     return make_iterator(std::begin(value), std::end(value), extra...);
 }
 
-template <typename InputType, typename OutputType> void implicitly_convertible() {
-    auto implicit_caster = [](PyObject *obj, PyTypeObject *type) -> PyObject * {
-        if (!detail::type_caster<InputType>().load(obj, false))
-            return nullptr;
-        tuple args(1);
-        args[0] = obj;
-        PyObject *result = PyObject_Call((PyObject *) type, args.ptr(), nullptr);
-        if (result == nullptr)
-            PyErr_Clear();
-        return result;
-    };
-    auto &registered_types = detail::get_internals().registered_types_cpp;
-    auto it = registered_types.find(std::type_index(typeid(OutputType)));
-    if (it == registered_types.end())
-        pybind11_fail("implicitly_convertible: Unable to find type " + type_id<OutputType>());
-    ((detail::type_info *) it->second)->implicit_conversions.push_back(implicit_caster);
+NAMESPACE_BEGIN(detail)
+
+template <typename InputType, typename OutputType>
+void* instance_create(void *input) {
+    // Takes a pointer to the InputType object, allocates an OutputType with the implicit
+    // conversion value and moves/copies the converted value into it, returning a pointer.
+    // Returns nullptr if input is nullptr.  The caller is responsible for destruction.
+    if (!input) return nullptr;
+    // The following is tempting, but will invoke *explicit* conversion paths, which we
+    // don't want to invoke:
+    // return new OutputType((OutputType) *reinterpret_cast<InputType*>(input));
+    // Instead we do this in two steps:
+    OutputType t = *reinterpret_cast<InputType*>(input); // implicit conversion
+    return new OutputType(std::move(t)); // move or copy
+}
+template <typename T>
+void instance_destroy(void *instance) {
+    delete (T*) instance;
+}
+// These have to be separated like this because the lambda in the proper version won't compile if
+// the C++ types aren't implicitly convertible.
+template <typename InputType, typename OutputType>
+typename std::enable_if<std::is_convertible<InputType, OutputType>::value>::type
+implicitly_cpp_convertible(detail::type_info &input_type) {
+    // Some validity checks on output type:
+    if (!std::is_destructible<OutputType>::value)
+        pybind11_fail("implicitly_convertible: " + type_id<OutputType>() + " is not destructible");
+    if (!std::is_move_constructible<OutputType>::value && !std::is_copy_constructible<OutputType>::value)
+        pybind11_fail("implicitly_convertible: " + type_id<OutputType>() + " is not move- or copy-constructible");
+
+    std::type_index outidx(typeid(OutputType));
+    auto &internals = get_internals();
+    auto &implicit_converters = internals.implicit_conversions_cpp[outidx];
+
+    // If the input type is a subclass of something else in the implicit conversion list, we need to
+    // put this one in just before it (so that this gets found first); otherwise we add to the end
```

Comment:
Is this comment in sync with the code below? To me, it seems that it it's not a subtype of any element in the list, the while loop will immediately terminate. Afterwards, the element will be inserted at the beginning and not the end of the list.

Related question: is this logic being tested by some part of your tests?